components/object_store/src/obkv/mod.rs

```diff
@@ -0,0 +1,960 @@
+// Copyright 2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    collections::HashSet,
+    hash::{Hash, Hasher},
+    ops::Range,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+    time,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use async_trait::async_trait;
+use bytes::Bytes;
+use chrono::{DateTime, TimeZone, Utc};
+use common_util::error::{BoxError, GenericError};
+use futures::{
+    stream::{BoxStream, FuturesOrdered},
+    StreamExt,
+};
+use log::debug;
+use snafu::{ensure, ResultExt, Snafu};
+use table_kv::{ScanContext, ScanIter, TableKv, WriteBatch, WriteContext};
+use tokio::{
+    io::{AsyncWrite, AsyncWriteExt},
+    time::Instant,
+};
+use twox_hash::XxHash64;
+use upstream::{
+    path::{Path, DELIMITER},
+    Error as StoreError, GetResult, ListResult, MultipartId, ObjectMeta, ObjectStore, Result,
+};
+
+use crate::{
+    multipart::{CloudMultiPartUpload, CloudMultiPartUploadImpl, UploadPart},
+    obkv::meta::{MetaManager, ObkvObjectMeta, OBJECT_STORE_META},
+};
+
+mod meta;
+mod util;
+
+/// The object store type of obkv
+pub const OBKV: &str = "OBKV";
+
+/// Hash seed to build hasher. Modify the seed will result in different route
+/// result!
+const HASH_SEED: u64 = 0;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Failed to scan data, namespace:{namespace}, err:{source}"))]
+    ScanData {
+        namespace: String,
+        source: GenericError,
+    },
+
+    #[snafu(display("Failed to put data, path:{path}, err:{source}"))]
+    PutData { path: String, source: GenericError },
+
+    #[snafu(display("Failed to create shard table, table_name:{table_name}, err:{source}"))]
+    CreateShardTable {
+        table_name: String,
+        source: GenericError,
+    },
+
+    #[snafu(display("Failed to read meta, path:{path}, err:{source}"))]
+    ReadMeta { path: String, source: GenericError },
+
+    #[snafu(display("None data found in part_key:{part_key}"))]
+    NoneDataPart { part_key: String },
```

Comment:
NoneDataPart is wired, how about `DataPartNotFound`?
```suggestion
    #[snafu(display("Data part is not found, part_key:{part_key}.\nBacktrace:\n{backtrace}"))]
    DataPartNotFound { part_key: String, backtrace: Backtrace },
```

BTW, the backtrace seems missing for other errors, maybe we should attach it to these errors.

Reply:
Done.